tests/integration/test_charm.py

```diff
@@ -3,29 +3,49 @@
 # See LICENSE file for licensing details.
 
 import logging
-from pathlib import Path
 
 import pytest
-import yaml
 from pytest_operator.plugin import OpsTest
 
+from .helpers import (
+    APP_NAME,
+    get_cluster_endpoints,
+    get_cluster_members,
+    get_juju_leader_unit_name,
+    get_key,
+    put_key,
+)
+
 logger = logging.getLogger(__name__)
 
-METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
-APP_NAME = METADATA["name"]
+NUM_UNITS = 3
 
 
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
 @pytest.mark.group(1)
 @pytest.mark.abort_on_fail
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
-    """Build the charm-under-test and deploy it together with related charms.
+    """Build the charm-under-test and deploy it with three units.
 
-    Assert on the status before any relations/configurations take place.
+    The initial cluster should be formed and accessible.
     """
     # Build and deploy charm from local source folder
     etcd_charm = await ops_test.build_charm(".")
+    model = ops_test.model_full_name
 
     # Deploy the charm and wait for active/idle status
-    await ops_test.model.deploy(etcd_charm, num_units=1)
-
+    await ops_test.model.deploy(etcd_charm, num_units=NUM_UNITS)
     await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+
+    # check if all units have been added to the cluster
+    endpoints = get_cluster_endpoints(ops_test, APP_NAME)
+    leader_unit = await get_juju_leader_unit_name(ops_test, APP_NAME)
+
+    cluster_members = get_cluster_members(model, leader_unit, endpoints)
+    assert len(cluster_members) == NUM_UNITS
+
+    # make sure data can be written to the cluster
+    test_key = "test_key"
+    test_value = "42"
+    assert put_key(model, leader_unit, endpoints, key=test_key, value=test_value) == "OK"
+    assert get_key(model, leader_unit, endpoints, key=test_key) == test_value
```

Comment:
I would check all 3 units separately to ensure it was replicated.

Reply:
May be premature, and will be addressed more in-depth during HA - but that's a valid point. 
For HA, we also need to have continuous writes routines implemented as part of the tests. Which should have replication checks happen as part of the continuous writes interface.

Reply:
Agree and will be done when working on HA.